Nav Block: adds basic items justification controls

getdave opened the pull request:
Closes https://github.com/WordPress/gutenberg/issues/18557.
Closes #18898.

Enables the ability to _justify_ the items within the Nav Block to the:

* left (technically `flex-start`)
* center
* right (technically `flex-end`)

This is not the same as _alignment_ controls which actually [shift the _entire_ Block](https://github.com/WordPress/gutenberg/issues/18557#issuecomment-561214391). This approach was avoided because:

1. When `left` or `right` alignment was applied the _whole_ Block became "full width".
2. It was not possible to have a block which was _both_ `full-wide` and with the items aligned to the left/right.

The approach in this PR works around that by adding a custom _justification_ control which controls the alignment of individual _items_ (taken as a whole) rather than the Block as a unit.

## Questions

* Do we need custom icons for item justification or are the text justification ones we're using ok?
* Shall we add further flexbox layout options for `space-around`, `space-between`...etc? I'm tempted to say we save that for a future PR and just get the basics merged.

## How has this been tested?

Before testing/reviewing please read [the background discussion in the Issue](https://github.com/WordPress/gutenberg/issues/18557) to understand the motivation behind the justification approach.

This has been manually tested. 

1. Add Nav Block.
2. Add lots of items.
3. Use controls to justify the items.
4. Use alignment to modify the alignment widths whilst also toggling the item justification. 
5. Check visual output corresponds correctly in both the **Editor** _and_ on the **front of site** (Theme).

### Deprecation Testing

This PR doesn't need to add a deprecation because we don't change the output of the Block until the user selects a justification option (ie: the default is don't change anything).

You can test this by

1. Switch to `master`.
2. Add Post with Nav Block with items.
3. Switch to this branch.
4. Reload your Page from `#2`.
5. Ensure the Block does not error.

## Screenshots <!-- if applicable -->
![Screen Capture on 2019-12-04 at 12-41-01](https://user-images.githubusercontent.com/444434/70143446-8a07ec00-1693-11ea-9890-cf72a4b22fcc.gif)

## Types of changes
New feature (non-breaking change which adds functionality)

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->.

getdave committed:
Adds basic item justification controls

getdave committed:
Update terminology to “Justify”

getdave committed:
Fixes i18n

getdave committed:
Adds Front end rendering of justification setting

jasmussen commented:
Thank you for the clarification. Alignment controls don't seem to add a ton of value given how the horizontal width of the block can grow. But justification controls _do_ make sense. You might create a 2 column layout with a logo in the left column, a nav block in the right column, and then wanting to justify the items inside the nav block to stack towards the right. 

In that vein, 👍 👍 on the feature. 

I wonder where in the toolbar the alignments should sit, they are almost never at the end. It should at least sit before colors, should it also sit before block navigator? I'd tend to think so. The overall pattern hasn't been fully established, but usually it goes like this

```
[switcher] [alignments] [inline] [plugins] [ellipsis]
```

So starting with block level first and moving towards inline and "other".

getdave committed:
Remove default to avoid need for Block deprecation

getdave committed:
Updates order of justification controls in toolbar

Addreses https://github.com/WordPress/gutenberg/pull/18909#issuecomment-561635165

getdave commented:
@jasmussen Thank you for the 👍 on the feature.

I've updated the toolbar order. Good spot!

getdave committed:
Fix potential PHP linting issue with trialing comma

ZebulanStanphill commented on packages/block-library/src/navigation/edit.js:
```diff
@@ -84,6 +86,14 @@ function Navigation( {
 	//
 	// HANDLERS
 	//
+	const handleItemsAlignment = ( align ) => {
```
For the sake of better debugging (since anonymous functions like this one lack names in stack traces), I think this should be changed to a standard function declaration. Writing it as an anonymous function assigned to a constant variable doesn't save any characters anyway.

There are other anonymous functions that could be changed to function declarations below this one, but I'll leave it up to you whether to change those in this PR or not.

getdave replied:
I understand. I see this pattern a lot across the codebase so I'd be interested to hear @youknowriad's opinion on whether it should be a code style thing to avoid anon functions. 

I've converted them for now.

ZebulanStanphill commented on packages/block-library/src/navigation/style.scss:
```diff
@@ -142,6 +142,23 @@
 		}
 	}
 
+	&.items-justification-left {
```
The nesting for the `> ul` rules is unnecessary since you can just do &.items-justification-left > ul`. Same for the next couple of rules.

getdave replied:
Yeh not sure why I did that. I'm breaking my own rules doing that! 

Thanks. All fixed.

retrofox commented on packages/block-library/src/navigation/index.php:
```diff
@@ -15,7 +15,7 @@
 	// CSS classes.
 	$colors = array(
-		'css_classes'   => [],
+		'css_classes'   => array(),
```
just curious about this change.

getdave replied:
It's automatically applied by WP Coding Standards linter.

ZebulanStanphill commented on packages/block-library/src/navigation/edit.js:
```diff
@@ -26,6 +27,7 @@ import {
 	Placeholder,
 	Spinner,
 	ToolbarGroup,
+	Toolbar,
```
Nitpick: alphabetically, `Toolbar` should come before `ToolbarGroup`. The rest of the imports in this list are listed alphabetically.

ZebulanStanphill reviewed (changes requested):
Haven't actually tested the code yet, but I noticed some minor things that could be improved.

retrofox commented:
I don't see the `wide-width`, `full-width` toolbar in the current PR. Are we going to add (another) alignment toolbar? ( I see both in your gif animation)
These are mostly layout options rather than alignment ones. Anyway, I wouldn't like to bring and open this topic (again 😅 )

if we are not going to add two toolbars and just adds only one, could extend the current navigation adding these justify options? I've copied/pasted what you wrote straight in the navigation block.

**packages/block-editor/src/components/block-alignment-toolbar/index.js**
```es6
const BLOCK_ALIGNMENTS_CONTROLS = {
	// ...,
	justifyLeft: {
		icon: 'editor-alignleft',
		title: __( 'Justify items left' ),
	},
	justifyCenter: {
		icon: 'editor-aligncenter',
		title: __( 'Justify items center' ),
	},
	justifyRight: {
		icon: 'editor-alignright',
		title: __( 'Justify items right' ),
	},
};
```

**packages/block-library/src/navigation/index.js**

```es6
export const settings = {
	// ...,

	supports: {
		align: [ 'justifyLeft', 'justifyCenter', 'justifyRight' ],
		// ...,
	},

	// ...,
};
```

After that, it's a matter of defining the CSS on both sides.

Sorry if this suggestion changes considerably this approach, but I think it worth because it's simpler since re-using blocks.

mtias commented on packages/block-library/src/navigation/edit.js:
```diff
@@ -148,17 +158,32 @@ function Navigation( {
+					label={ __( 'Change items justification' ) }
+					isCollapsed
+					controls={ [
+						{ icon: 'editor-alignleft', title: __( 'Justify items left' ), isActive: 'left' === attributes.itemsJustification, onClick: handleItemsAlignment( 'left' ) },
```
We could consider using a different set of icons, closer to what we use for columns vertical alignment. cc @jasmussen 

![image](https://user-images.githubusercontent.com/548849/70171238-98bfc480-16ce-11ea-9540-4a53132778a5.png)

jasmussen replied:
Yes that's an excellent point, as you're not aligning _text_.

getdave replied:
I think we should but could someone make some for me please? 😁

jasmussen replied:
Can you reuse the ones from the columns block?

![Screenshot 2019-12-04 at 20 29 09](https://user-images.githubusercontent.com/1204802/70174347-ff47e100-16d4-11ea-89a7-6e20670ccd4d.png)

or wait, maybe you mean horizontal versions, right?

I'll get them here for you tomorrow if that's okay.

jasmussen replied:
Here you go. Left:

```
<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M11 16v-3h10v-2H11V8l-4 4 4 4zM5 4H3v16h2V4z"/></svg>
````

Center:

```
<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M5 8v3H1v2h4v3l4-4-4-4zm14 8v-3h4v-2h-4V8l-4 4 4 4zM13 4h-2v16h2V4z"/></svg>
```

Right:
```
<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M13 8v3H3v2h10v3l4-4-4-4zm8-4h-2v16h2V4z"/></svg>
```

getdave replied:
Super helpful! Thank you

<img width="499" alt="Screen Shot 2019-12-05 at 09 09 34" src="https://user-images.githubusercontent.com/444434/70220777-04da1100-173f-11ea-8d2e-eed9f5e5e456.png">

getdave committed:
Adds custom icons

Provided here https://github.com/WordPress/gutenberg/pull/18909#discussion_r354145223

getdave committed:
Fix unecessary CSS nesting

Addresses https://github.com/WordPress/gutenberg/pull/18909#discussion_r353826291

getdave committed:
Fix to optimise components imports alphabetically

Addresses https://github.com/WordPress/gutenberg/pull/18909#discussion_r353828813

getdave committed:
Update anon funcs to named funcs to aid debugging

Resolves https://github.com/WordPress/gutenberg/pull/18909#discussion_r353823644

getdave commented:
@retrofox Thanks for your comments. Allow me to address them.

> I don't see the wide-width, full-width toolbar in the current PR. 

The alignment toolbar is not in the `edit.js` because it gets added automatically via the `supports.align` config in `index.js`. You don't have to code it manually.

> Are we going to add (another) alignment toolbar?

In this block yes there are two things that control placement on the hoz axis:

1. Alignment - this weirdly controls the width even though it's called "alignment". 
2. Justification - this is distinguished from "alignment" because it's justifiying the _items_ on the hoz axis not the Block itself. 🤷‍♂ 

I think this might be a use case specific to Nav. I'm not sure. 

> ...if we are not going to add two toolbars and just adds only one, could extend the current navigation adding these justify options? 

No because of the reasons outlined in the PR description, namely:

* If you have a single toolbar you can't select more than 1 option at a time. You might want full width + justified right for example.
* Alignment and Jusitifcation are not the same thing and colocating them under 1 toolbar could be confusing especially if `alignLeft/Right/Center` was added. 

My feeling at this stage is that I'm keen not to over abstract justification controls into a reusable toolbar too soon. I don't yet know of any other use cases where this functionality would be required. Moreover, this feature needs to be land in the Nav Block _asap_. 

Going off down a reusable route will introduce additional complexity. I'd be happy to handle that in a follow-up PR but my instinct is that it's premature to address it here.

As always, I'm open to being told I'm wrong! 😄

getdave commented:
@jasmussen @mtias We now have the new icons in place.

getdave committed:
Fixes to output align wide/full classes

getdave committed:
Updates classname to proper English adjective form

mtias commented:
It doesn't seem to be affecting the display:

![image](https://user-images.githubusercontent.com/548849/70229419-7a051080-1756-11ea-907b-119b0a4b58d5.png)

getdave committed:
Fix to ensure justify works for Nav with few items

Previously the nav inner elements didn’t occupy the full space of their container. As a result whilst the elements were justified it didn’t appear so because the containing element was only as wide as the width of the items.

getdave commented:
> It doesn't seem to be affecting the display:
> 
> ![image](https://user-images.githubusercontent.com/548849/70229419-7a051080-1756-11ea-907b-119b0a4b58d5.png)

@mtias This caught me out. Apologies. I was stress testing with a large nav with lots of items. This caused me to miss the issue.

The issue was the nav inner containing elements didn’t occupy the full space of their container. As a result whilst the elements were justified it didn’t appear so because the containing element was only as wide as the width of the items.

This is fixed.

<img width="708" alt="Screen Shot 2019-12-05 at 11 42 10" src="https://user-images.githubusercontent.com/444434/70232418-504aea00-1754-11ea-9407-ed5a8eb310b1.png">

draganescu reviewed (commented):
I tested and it worked very well!

getdave commented:
@draganescu Thanks for your review 👍 

@ZebulanStanphill Please could you review your review as it's now blocking merge? Much appreciated.

ZebulanStanphill reviewed (approved):
Everything seems to be working as expected. :+1:

obenland reviewed (approved):
Nice work @getdave!

shaunandrews commented:
I saw the new "Justify" settings for the first time today and was completely confused. It feels like we're making a distinction between "justifying" and "aligning" for reasons that most people will never understand. I've never seen the word "justify" used like this before, and would not know what it meant. "Align" would be way more obvious to me.

The use of new icons is also introducing more overhead. These icons don't help me understand that I can align my navigation items to the left or right. To me, they look more indent.

ZebulanStanphill commented:
@shaunandrews I think part of the issue is that there is already a block alignment toolbar, but it doesn't really contain any actual alignments except for `aligncenter`. `alignleft` and `alignright` are floats, and the wide and full options similarly have nothing to do with "alignment". The options really feel more like "flow" options that affect how the block is positioned in the document.
